credentials/sts/sts.go

```diff
@@ -0,0 +1,427 @@
+		SubjectToken:       string(subToken),
+		SubjectTokenType:   c.opts.SubjectTokenType,
+	}
+	actorToken, err := readActorTokenFrom(c.opts.ActorTokenPath)
+	if err == nil {
+		reqParams.ActorToken = string(actorToken)
+		reqParams.ActorTokenType = c.opts.ActorTokenType
+	}
+	jsonBody, err := json.Marshal(reqParams)
+	if err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequest("POST", c.opts.TokenExchangeServiceURI, bytes.NewBuffer(jsonBody))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create http request: %v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	return req, nil
+}
+
+// sendRequestWithRetry sends the provided http.Request and retries with
+// exponential backoff for certain types of errors. It takes care of closing the
+// response body on success (returns the contents of the body) and retries.
+func (c *callCreds) sendRequestWithRetry(ctx context.Context, req *http.Request) ([]byte, error) {
+	retries := 0
+	for {
+		// Wait for one of exponential backoff or context deadline to expire.
+		if retries > 0 {
+			timer := time.NewTimer(c.bs.Backoff(retries))
+			select {
+			case <-ctx.Done():
+				timer.Stop()
+				return nil, ctx.Err()
+			case <-timer.C:
+			}
+		}
+
+		// http.Client returns a non-nil error only if it encounters an error
+		// caused by client policy (such as CheckRedirect), or failure to speak
+		// HTTP (such as a network connectivity problem). A non-2xx status code
+		// doesn't cause an error.
+		resp, err := c.client.Do(req)
+		if err != nil {
+			return nil, err
+		}
+
+		// When the http.Client returns a non-nil error, it is the
+		// responsibility of the caller to read the response body till an EOF is
+		// encountered and to close it.
+		body, err := ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
+		if err != nil {
+			retries++
+			continue
+		}
+
+		if resp.StatusCode == http.StatusOK {
+			return body, nil
+		}
+		if resp.StatusCode >= http.StatusBadRequest && resp.StatusCode < http.StatusInternalServerError {
+			// For 4xx errors, which are client errors, we do not retry.
+			return nil, fmt.Errorf("http status %d, body: %s", resp.StatusCode, string(body))
+		}
+		retries++
+		grpclog.Warningf("http status %d, body: %s", resp.StatusCode, string(body))
+	}
+}
+
+func tokenInfoFromResponse(respBody []byte) (*tokenInfo, error) {
+	respData := &ResponseParameters{}
+	if err := json.Unmarshal(respBody, respData); err != nil {
+		return nil, fmt.Errorf("json.Unmarshal(%v): %v", respBody, err)
+	}
+	if respData.AccessToken == "" {
+		return nil, fmt.Errorf("empty accessToken in response (%v)", string(respBody))
+	}
+	return &tokenInfo{
+		tokenType:  respData.TokenType,
+		token:      respData.AccessToken,
+		expiryTime: time.Now().Add(time.Duration(respData.ExpiresIn) * time.Second),
+	}, nil
+}
+
+// RequireTransportSecurity indicates whether the credentials requires
+// transport security.
+func (c *callCreds) RequireTransportSecurity() bool {
```

Comment:
FYI, the "modern" version of this is to call `credentials.CheckSecurityLevel` in `GetRequestMetadata`: https://godoc.org/google.golang.org/grpc/credentials#CheckSecurityLevel

This should be used to distinguish between integrity-only and integrity+privacy, if the distinction is important for these credentials (which I imagine it is?).